### PR TITLE
Nuclia understanding api

### PR DESCRIPTION
**Pull request by ebrehault**
# Description

This pull request allows to use the [Nuclia](https://docs.nuclia.dev/docs/about/nuclia) file processing capabilities in LangChain.

It provides:

- a tool allowing to push files to the Nuclia processing queue, and to pull results fron this queue.
- a document loader wrapping the tool, to ease the usage and to comply with the LangChain practices.

(both comes with an example notebook + unit tests)

That's mostly about extracting text and metadata from any kind of file (document or media). It does *not* provide the other services offered by the Nuclia API.

I will do other PRs to cover:
- entity recognition
- vector storage and searching in the NucliaDB vector database
(unless you prefer a single PR providing all the different services altogether, let me know)

# Dependencies

It requires an up-to-date version of `protobuf` (which is already a dependency of LangChain) and the `nucliadb-protos` package, in order to parse Protobuf messages retrieved from the Nuclia API.

@rlancemartin, @eyurtsev,  @hinthornw, please review it when you have a moment :)
And thank you for your work on LangChain, that's great!

Note: our Twitter handler is `@NucliaAI`

### CODE REVIEW

File: langchain/document_loaders/nuclia.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from langchain.tools.nuclia.tool import NucliaUnderstandingAPI
+
+
+class NucliaLoader(BaseLoader):
+    """Loader that loads Nuclia Understanding API results."""
+
+    def __init__(self, path: str, nuclia_tool: NucliaUnderstandingAPI):
+        self.nua = nuclia_tool
+        self.id = str(uuid.uuid4())
+        self.nua.run(
```
Comment (baskaryan) [inline langchain/document_loaders/nuclia.py:17]: would it be cleaner to store path here and then run this command in L23 with self.path?

File: langchain/document_loaders/nuclia.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    def load(self) -> List[Document]:
+        """Load documents."""
+        data = self.nua.run(
+            {"action": "pull", "id": self.id, "path": None, "enable_ml": False}
+        )
+        if not data:
+            return []
+        obj = json.loads(data)
+        text = obj["extracted_text"][0]["body"]["text"]
```
Comment (baskaryan) [inline langchain/document_loaders/nuclia.py:29]: out of curiosity, why don't we iterate overall all extracted texts and return multiple documents?

File: langchain/tools/nuclia/tool.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    def _run(
+        self,
+        action: str,
+        id: str,
+        enable_ml: bool,
+        path: Optional[str],
+        run_manager: Optional[CallbackManagerForToolRun] = None,
+    ) -> str:
+        """Use the tool."""
+        zone = os.environ.get("NUCLIA_ZONE", "europe-1")
```
Comment (baskaryan) [inline langchain/tools/nuclia/tool.py:74]: could we pass zone and nua_key in during instantiation?

**Review by baskaryan — COMMENTED** [review]
very cool! left a few questions

File: langchain/document_loaders/nuclia.py (hunk 1 above)
Comment (ebrehault) [reply langchain/document_loaders/nuclia.py:17]: Actually we need the `path` value only here (line 18).
On line 23, we do `"path": None`.

File: langchain/document_loaders/nuclia.py (hunk 2 above)
Comment (ebrehault) [reply langchain/document_loaders/nuclia.py:29]: At the moment, the data structure returns a list but it always contain only one item.
That's something we will fix in the future.

File: langchain/tools/nuclia/tool.py (hunk 3 above)
Comment (ebrehault) [reply langchain/tools/nuclia/tool.py]: right, fixed

**Review by rlancemartin — COMMENTED** [review]
Nice! Great contribution. 

Should we add this as a `document transformer`? 

The majority of the processing appears to be generically applicable to any `Document`:

```
file metadata
extracted text
nested text (like text in an embedded image)
a summary (only when enable_ml is set to True)
paragraphs and sentences splitting (defined by the position of their first and last characters, plus start time and end time for a video or audio file)
named entities: people, dates, places, organizations, etc. (only when enable_ml is set to True)
links
embedded files
the vector representations of the text (only when enable_ml is set to True)
```

Adding as a `transformer` to [this directory](https://github.com/hwchase17/langchain/tree/master/langchain/document_transformers) is a fairly small change. 

It can then be applied to any document (from any loader):
```
documents = [Document(page_content=sample_text)]
nuclia_transformer = NucliaTransformer()
transformed_document = nuclia_transformer.transform_documents(documents)
```

**Comment by ebrehault** [comment]
@rlancemartin Thank you for your feedback!

> Should we add this as a document transformer?

Yes definitely. So I have just pushed a document transformer implementation.

I kept the document loader though, because the Nuclia API can process plain text (hence the document transformer) but also files (like .pdf, .docx, .mp4, or any type you want), so I guess a document loader also makes sense here.

Regarding the document transformer implementation, considering the Nuclia Understanding API works with a processing queue, it is totally asynchronous, so I figured it was easier to implement the transformer as async using the `atransform_documents` (while `transform_documents` is not implemented as it would be probably particularly ugly and would provide a bad developer experience). I hope that's fine.

Consequently, the Nuclia tool now supports both sync and async modes, which could be handy in some cases.

**Comment by rlancemartin** [comment]
> @rlancemartin Thank you for your feedback!
> 
> > Should we add this as a document transformer?
> 
> Yes definitely. So I have just pushed a document transformer implementation.
> 
> I kept the document loader though, because the Nuclia API can process plain text (hence the document transformer) but also files (like .pdf, .docx, .mp4, or any type you want), so I guess a document loader also makes sense here.
> 
> Regarding the document transformer implementation, considering the Nuclia Understanding API works with a processing queue, it is totally asynchronous, so I figured it was easier to implement the transformer as async using the `atransform_documents` (while `transform_documents` is not implemented as it would be probably particularly ugly and would provide a bad developer experience). I hope that's fine.
> 
> Consequently, the Nuclia tool now supports both sync and async modes, which could be handy in some cases.

Great! I just kicked off tests.

**Comment by rlancemartin** [comment]
There is a problem w/ `poetry.lock` - 
```
Error: poetry.lock is not consistent with pyproject.toml. Run `poetry lock [--no-update]` to fix it.
```

Run this to re-generate the lock file -
```
git checkout master
git pull
git checkout BRANCH
git merge master
git checkout master poetry.lock
< add your updates to pyproject.toml >
poetry lock --no-update
```

**Comment by ebrehault** [comment]
@rlancemartin just a question: where I am supposed to declare my dependency to `nucliadb_protos` so it gets install when running tests in GitHub Actions?

**Comment by ebrehault** [comment]
Hi @rlancemartin!
The PR has been cleaned and rebased on the recent changes done on the repository structure.
Could you please make a final review, and hopefully it can be merged before I get a new conflict on `poetry.lock` :)

**Comment by baskaryan** [comment]
landed in #8404, thank you @ebrehault for the contribution!!

**Pull request closed without merging**